### DISCUSSION TITLE
Metadata-based Automatic Parameter Binding

### DISCUSSION THREAD

**Discussion by AdamSobieski** [dotnet/csharplang#6846] [Language Ideas]
Hello. I would like to share some ideas pertaining to metadata-based automatic parameter binding.

While I'm exploring these topics in the context of [a new language](https://github.com/w3c-cg/planning/discussions/28) and mostly with respect to `<>`-based template arguments, I’m hoping that these ideas might be useful for some C#-related brainstorming.

The idea is that developers could utilize metadata to enable automaticity with respect to binding values from containing scope(s) when invoking methods.

Metadata on method parameters could be used to specify when and which values from containing scope(s) could be bound when arguments are not manually typed.

In these regards, there are a number of approaches to consider.

One approach is to utilize URI-based semantics to indicate when a parameter to argument match exists:
```cs
public void foo([Provides("http://www.uri.org/#semantic")] int x)
{
    baz(); // can be invoked in this scope without the explicit argument as the scope contains a matching provider
}
public void baz([Accepts("http://www.uri.org/#semantic")] int x)
{
    ...
}
```

A second approach is to utilize, instead of matching URIs, matching `System.Type`s: 
```cs
public void foo([Provides(typeof(Example.SlotA))] int x)
{
    baz(); // can be invoked in this scope without the explicit argument as the scope contains a matching provider
}
public void baz([Accepts(typeof(Example.SlotA))] int x)
{
    ...
}
```

A third approach involves utilizing something like `System.Reflection.Binder` which receives all of the candidate values from the containing scope(s), and provides the to-be-invoked method with one:
```cs
public void foo([SomeAttribute] int x)
{
    baz(); // can be invoked in this scope without the explicit argument as the scope contains a matching provider
}
public void baz([Binder(typeof(Example.MyBinder))] int x)
{
    ...
}
```

A fourth approach requires first-class rules in the language. Binding would be able to occur via the utilization of pattern-matching rules on the set of candidate values from containing scope(s). It might be possible to implement rule-based approaches for C# with classes implementing certain interfaces.

Interestingly, there might be yet other solutions and approaches. With any solution, developers would not have to type arguments when they could be inferred in context. The examples, above, were simplified for clarity and these solutions would be more useful for scenarios with dozens of variables in scope and with dozens of method invocations to type.

These techniques could also be useful for scenarios where metadata-based binding verification would be convenient, e.g., when all of the variables in a scope were of type `System.Object` or were all `dynamic`.

These techniques could also be useful for advancing IntelliSense. When developers typed the opening parenthesis for a method invocation, inference could provide a suggestion. Developers could press the `tab` key and then use arrow keys to iterate through possibilities, in the event that more than one binding of values was possible for a method invocation.

These metadata-based techniques could also be useful for enhancing inferencing capabilities with respect to generic arguments for generic classes and methods.

In integrated development environments, text coloring could be utilized to visually indicate when a manually, explicitly typed arguments could have been omitted, implicit, inferred in the context. Compiler warnings could additionally be produced when typed method invocations were contrary to their metadata descriptions.

**Comment by CyrusNajmabadi** [comment]
> One approach is to utilize URI-based semantics to indicate when a parameter to argument match exists:

Can you explain what "URI-based semantics" means here?

**Reply by AdamSobieski** [reply]
If one were to use strings for defining when a match signified that a provided value matched a parameter for an automatic or implicit binding, URIs would be useful to utilize as such across assemblies and projects (see also: [Semantic Web](https://en.wikipedia.org/wiki/Semantic_Web)). Other options would include period-delimited strings like `namespace.subnamespace.slot`.

**Comment by CyrusNajmabadi** [comment]
> The examples, above, were simplified for clarity and these solutions would be more useful for scenarios with dozens of variables in scope and with dozens of method invocations to type.

This does not seem like a virtue.  If you have dozens of variables in scope, having it be unclear about which were passed in definitely seems like a recipe for confusion.

The above proposal doesn't seem to actually show examples of where this would be useful or why this would be desirable.

**Comment by CyrusNajmabadi** [comment]
@AdamSobieski Can you explain what the difference would be between this and the championed proposal for implicit parameters here: https://github.com/dotnet/csharplang/issues/6300

**Reply by AdamSobieski** [reply]
A difference appears to be that this proposal points to a more expressive metadata-based approach where #6300 wants to introduce a new keyword to provide the same or similar features. It could be that metadata-based approaches could complement a new `implicit` keyword, enhancing the inferencing capabilities.

**Reply by CyrusNajmabadi** [reply]
>  to a more expressive metadata-based approach

Why is this valuable or desirable?  What's the use case where this is needed?

@AdamSobieski i've noticed with a lot of your discussions that tehre appears to be no use cases actually presented.  How would these ideas, for example, improve the .Net platform or the BCL?  How would they end up benefiting existing users with their apps and libraries?

> enhancing the inferencing capabilities.

What are the motivating scenarios for this?

**Reply by AdamSobieski** [reply]
My motivating scenarios are from [another programming language](https://github.com/w3c-cg/planning/discussions/28) which I am working on towards equipping developers for AI-related tasks such as planning-based story generation and natural-language generation.

The language resembles:

```
public typeofthing identifier<[attr1] ?x, [attr2] ?y, [attr3] ?z>
{
    ...
}
```

where types of things include first-class: predicates, rules, sets, knowledgebases, queries, operators, classes, properties, attributes / decorators, and templates.

Developers will often reference things with `<>` arguments from in-scope variables, and I desired for developers to only have to type the `identifier` portions and to be able to omit those arguments in the `<>` brackets which could be automatically inferred for them. These metadata-related ideas originally occurred to enable developers to be able to reference language constructs, like .NET's classes and interfaces, with `<>` parameters, without having to explicitly provide the `<>` arguments.

I then considered that these ideas could also be applicable to method invocation and that they might be generally useful for C#/.NET . It looks like some of these method-invocation-related topics were already discussed in #6300 .

**Reply by CyrusNajmabadi** [reply]
> and I desired for developers to only have to type the identifier portions and to be able to omit those arguments in the <> brackets which could be automatically inferred for them. 

Why is this beneficial?  Ask the examples given so far have been extremely vague. Having users explicitly pass these parameters seems extremely clear and understandable. 

It also seems like you're trying to make c# *extremely* dynamic, where things like method binding are not resolved by the compiler, but by some extension. First, that really goes against what c# wants to be as a highly static language. Second, you haven't provided information on why this would be a good thing. 

I don't see why "equipping developers for AI-related tasks such as planning-based story generation and natural-language generation" requires any of these things. These generation tasks seem just as possible with normal static typing and normal overload binding. 

When I ask for examples and use cases, I mean actual snippets of real code solving real problems that these features would then allow to become much better on some axis.  So that would involve being able to see the before/after for that code in that domain to make it clear why this would be a good thing to do.  Thanks!

**Reply by AdamSobieski** [reply]
Towards an example showing before and after, from the prototype language, let us consider creating and defining an intensional set which utilizes a number of rules in an and/or tree for its semantic definition.

Without template-argument inference, developers would have to type something like:

```
using system;

namespace example
{
    public rule rule123<[world] ?w, ?set, ?x>
    {
        ...
    }

    ...

    public set set123<[world] ?w, [self] ?s, [element] ?e>
    {
        contains
        {
            or
            {
                and
                {
                    rule123<?w, ?s, ?e>;
                    rule124<?w, ?s, ?e>;
                }
                and
                {
                    rule125<?w, ?s, ?e>;
                    rule126<?w, ?s, ?e>;
                }
            }
        }
    }
}
```

With metadata-enhanced template-argument inference, developers could instead type something like:

```
using system;

namespace example
{
    public rule rule123
    <
        [auto][world][bind(*, system.world)]        ?w,
        [auto][bind(set::contains, system.self)]    ?set,
        [auto][bind(set::contains, system.element)] ?x
    >
    {
        ...
    }

    ...

    public set set123<[world] ?w, [self] ?s, [element] ?e>
    {
        contains
        {
            or
            {
                and
                {
                    rule123;
                    rule124;
                }
                and
                {
                    rule125;
                    rule126;
                }
            }
        }
    }
}
```

Developers could choose to refer to rules, in this case, without having to repetitively type the template arguments from the scope context.

Also, since there are first-class rules, we could use pattern-matching rules as arguments to the attributes or decorators:

```
using system;
using w3c.rdf;

namespace example
{
    public rule pmr1<[world] ?w>
    {
        ...
    }

    public rule pmr2<[world] ?w>
    {
        ...
    }

    public rule pmr3<[world] ?w>
    {
        ...
    }

    public rule rule123
    <
        [auto][world][bind(*, pmr1)]      ?w,
        [auto][bind(set::contains, pmr2)] ?set,
        [auto][bind(set::contains, pmr3)] ?x
    >
    {
        preconditions
        {
            holds(?w, { type(?set, Set); });
        }
        consequent
        {
            holds(?w, { contains(?set, ?x); });
        }
        script
        {
            ...
        }
    }

    public set set123<[world] ?w, [self] ?s, [element] ?e>
    {
        contains
        {
            or
            {
                and
                {
                    rule123;
                    rule124;
                }
                and
                {
                    rule125;
                    rule126;
                }
            }
        }
    }
}
```

Hopefully, these two examples, above, show the improvement, developers wouldn't have to type `<?w, ?s, ?e>` for each utilized rule. There is a trade-off: more intricacy when defining things and then more convenience when reusing them.

These ideas originally intended to reduce typing for template-like things which often reference one another. It could be useful for C#/.NET with generic types and methods and for method invocation scenarios per #6300. The described features intend to convenience developers, but as you said, developers could still choose to type the explicit template or type parameters as necessary or convenient.

**Reply by CyrusNajmabadi** [reply]
@AdamSobieski I have no idea what the above represents.  It does not seem to be c# at all.

**Comment by CyrusNajmabadi** [comment]
>  It might be possible to implement rule-based approaches for C# with classes implementing certain interfaces.

I'm not sure what this means.  Are you just reimplementing types here?

**Reply by AdamSobieski** [reply]
I was thinking about ways that one could go about implementing (pattern-matching) rules in .NET. I meant to express that a class could implement something like `IRule` or `IPatternMatchingRule` (this could be a scenario for static abstract methods on interfaces). The IDE or compiler would load these types up from referenced assemblies, potentially activating or creating instances of them, and would use them for performing inferencing at development time. Pattern-matching rules, which can be implemented in a number of ways, could be more expressive than a type system as rules could utilize reflection, broadly, to compare candidate values for binding, this including usage of metadata.

**Reply by CyrusNajmabadi** [reply]
What ist he use case where this is valuable?

> could be more expressive than a type system as rules could utilize reflection, broadly, to compare candidate values for binding, this including usage of metadata.

Why is htis valuable or desirable?

**Reply by AdamSobieski** [reply]
Uses include:

1. automatically providing arguments for method invocations, automatically providing type arguments for generic classes and methods
    1. in some approaches the developer could omit the argument, in others, they would utilize a keyword like `auto` or `_`. With something in the method invocation position, the developer could mouseover to view the inference.
2. automatically producing better IntelliSense hints with enhanced inferencing.
    1. in these approaches, when the developer types `(` to begin typing the arguments, they would have available visual hints on screen.
    2. developers could type `(` and then `tab` to utilize enhanced inferencing without syntactic omission.
3. compiler informational messages and warnings could also be provided when developers invoke methods in ways contrary to the metadata-enhanced binding features.

Also, one use-case is where some parameter is repeatedly utilized throughout a portion of a software, like `globalThis` or a `RuntimeContext` instance:

```cs
public void example(RuntimeContext ctx, int x, int y, int z)
{
    foo(ctx, x, y);
    ...
    etc(ctx);
}

public void foo(RuntimeContext ctx, int a, int b)
{
    ...
}
public void etc(RuntimeContext ctx) { ... }
```
So, when some sorts of context variables are often utilized, passed to methods and provided to invoked methods, developers could use an `implicit` keyword and/or metadata to reduce typing.

That's one useful scenario; there are others to consider.

**Reply by CyrusNajmabadi** [reply]
> That's one useful scenario; there are others to consider.

That scenario looks to be already covered by the other proposal.  I'm asking what real world cases would benefit from the proposal you're making here.  Thanks! :)

**Reply by AdamSobieski** [reply]
I'm reading #6300 and it appears that the real-world scenarios for C#/.NET from that proposal are similar to those for this discussion post here. This discussion post introduces that metadata attributes could enhance implicit binding in general.

For example:
```
public void foo([Provides(typeof(SlotA))] int x, [Provides(typeof(SlotB))] int y)
{
    baz(); // here, two integers can simultaneously be passed to functions in this scope implicitly
}
public void baz([Accepts(typeof(SlotA))] int x, [Accepts(typeof(SlotB))] int y)
{
    ...
}
```

Types (from a secondary type model) like `SlotA` and `SlotB` could enhance implicit argument binding. Types could also be utilized to describe aspects, attributes, or features provided by values.

For example:
```
public void foo([Provides(typeof(FeatureA))] int x,
                [Provides(typeof(FeatureB))][Provides(typeof(FeatureC))] int y)
{
    baz();
}
```

If only we could use metadata attributes on local variable declarations in C#/.NET:

```
public void foo([Provides(typeof(FeatureA))] int x,
                [Provides(typeof(FeatureB))][Provides(typeof(FeatureC))] int y)
{
    [Provides(typeof(FeatureA))]
    int z = 123;

    baz();
}
```

**Reply by CyrusNajmabadi** [reply]
These examples are not helpful. 'foo' and 'baz' mean nothing here. I'm asking for real use cases where this capability is needed and beneficial.

**Reply by AdamSobieski** [reply]
I was hoping to show that parameters could be attributed or decorated with multiple `[Provide]` and multiple `[Accept]` attributes and that these metadata descriptions could enhance inferencing to provide automaticity and IntelliSense. As for more practical, specific, real-world applications of metadata-enhanced generic-argument inference and metadata-enhanced method-argument inference, perhaps you could propose one or more categories of development scenarios, like creating applications using ASP.NET, or some other concrete, yet broad, categories of scenarios and I could phrase examples in those categories.

**Reply by CyrusNajmabadi** [reply]
How about someone trying to work with a DB to populate information on a website?

**Reply by AdamSobieski** [reply]
```
public class Document { }
public class Username { }

public HtmlElement GenerateHtmlWidget
(
    [Implicit][Provides(typeof(Document))][Accepts(typeof(Document))] HtmlDocument document,
    [Provides(typeof(Username))][Accepts(typeof(Username))] string x,
    string y,
    string z
)
{
    var element = document.CreateElement(...);
    ...
    return element;
}
```

In the domain of generating hypertext from database data, there are a lot of methods which have a first parameter of type `HtmlDocument` which are utilized to create `HtmlElement`s, `HtmlAttribute`s, and so forth, per the Document Object Model. Furthermore, these methods are often nested, invoking one another, passing their first argument of type `HtmlDocument` to one another.

By utilizing `[Implicit]`, `[Provides(typeof(Document))]`, and `[Accepts(typeof(Document))]` as metadata on the first parameters of each of a set of methods, developers could omit passing the `HtmlDocument` object throughout that set of methods, this value being passed implicitly.

By providing metadata for the second parameter, `x`, developers could ensure that typographical errors do not result in other strings being provided for that parameter. Strings, for instance, can be used for many things and scenarios, and metadata could contribute to preventing typographical errors, enhancing type safety.

It is possible that: (1) a compiler could auto-populate the `HtmlDocument` argument, resembling the `implicit` keyword proposal (#6300), and that (2) a compiler could generate informational messages or warnings if arguments provided are not attributed with matching or compatible metadata.

**Reply by CyrusNajmabadi** [reply]
> By providing metadata for the second parameter, x, developers could ensure that typographical errors do not result in other strings being provided for that parameter.

I don't see why this is needed in a statically typed language.  Just have `Username x` instead of the vague `string x`.  Now the language ensures you're doing the right thing.

I feel like this is an attempt to graft dynamic semantics onto the language where there are already sufficient static options.

**Comment by 333fred** [comment]
This seems very similar to https://github.com/dotnet/csharplang/issues/6300.

Edit: And I missed that Cyrus already pointed that out.
